dynamic_validators/insecure_dynamic_validators.md

```diff
@@ -0,0 +1,128 @@
+
+
+## Proposed Change
+- Add a new subcommand `bighciandb upsert-validator PUBLIC_KEY POWER` which will allow the node admin to add/update/delete a validator.
+
+- `GET /api/v1/validators` should be introduced to list the current validators.
+
+### Technical challenges
+Since node admin can call `upsert-validator` anytime it is necessary to (temporarily) store the payload until `end_block` is called. This implies that just executing `bigchaindb upsert-validator` doesn't imply the validator is add to validators list.
+
+Furthermore, as discussed in [technical details](#technical-details) each/majority (>2/3) nodes in the network need to add the new node to their list of validators in order for the new node to be able to act as a validator. Until then the node may not be able to participate and vote on blocks.
+
+
+### Storing dynamically added validators
+In the previous section it was discussed that it is necessary to (temporarily) store the updates to validators list until `end_block` is executed. Following is one possible solution,
+
+- Store validator diff in a MongoDB collection along with a flag when `bighciandb upsert-validator` (discussed below) is called i.e.,
+```json
+{"validators": [{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+                 "power":10}],
+ "sync": true
+}
+```
+- The `"sync"` key is used during `end_block` to decide if a validator update should be returned.
+- During `end_block` all the documents with `"sync": true` are aggregated in order based on their insert time stamp and validator diff list is prepared which is returned in `end_blocks` response.
+
+NOTE: a validators diff which results in a change of `>1/3` of voting power will be rejected by tendermint.
+
+### API impact
+
+- A new subcommand `upsert-validator` should be introduced to add/update/delete validators
+  - syntax : `bichaindb upsert-validator PUBLIC_KEY POWER`
+  - example usage
+    - add/update validator
+    ```bash
+    $ bichaindb upsert-validator 4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A 10
+    ```
+    - delete validator
+    ```bash
+    $ bichaindb upsert-validator 4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A 0
+    ```
+
+- List validators
+  - Uri: `GET /api/v1/validators`
+  - response: 
+  ```json
+  [{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+    "power":10},
+   {"pub_key":{"type":"ed25519","data":"608D839D7100466D6BA6BE79C320F8B81DE93CFAA58CF9768CF921C6371F2553"},
+   "power":0}
+   ]
+  ```
+
+**NOTE**: when `power: 0` is set for a given validator it is removed from the validators list.
+
+### Workflow overview
+Below is a summary of workflow sequence which should be executed in order to add/remove a validator dynamically,
+- BigchainDB client wishes to add/remove validators from their own node.
```

Comment:
With BigchainDB, we tend to think of "clients" as being external users interacting with the BigchainDB network via the HTTP API only. We call the person operating a node the "node operator." (I know that other blockchain software uses the word "client" to mean a node in the network.) This line could be changed to:

`- BigchainDB node operator wishes to add/update/delete a validator from their own node.`